ipld/graphsync/src/response_manager/peer_response_sender.rs

```diff
@@ -0,0 +1,331 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::{LinkTracker, ResponseBuilder};
+use crate::{ExtensionData, GraphSyncResponse, RequestID, ResponseStatusCode, MAX_BLOCK_SIZE};
+use async_trait::async_trait;
+use cid::Cid;
+use libp2p::core::PeerId;
+
+/// Handles batching, deduping, and sending responses for a given peer across multiple requests.
+pub struct PeerResponseSender {
+    peer: PeerId,
+    link_tracker: LinkTracker,
+    response_builders: Vec<ResponseBuilder>,
+}
+
+impl PeerResponseSender {
+    /// Creates a new peer response sender.
+    pub fn new(peer: PeerId) -> Self {
+        Self {
+            peer,
+            link_tracker: LinkTracker::new(),
+            response_builders: Vec::new(),
+        }
+    }
+
+    /// Sends a given link for a given request ID across the wire, as well as its corresponding
+    /// block if the block is present and has not already been sent.
+    /// Returns true if the block has not already been sent and is thus added to a response.
+    pub fn send_response(&mut self, id: RequestID, link: Cid, data: Option<Vec<u8>>) -> bool {
+        let block_is_present = data.is_some();
+        let block_size = data.as_ref().map_or(0, |vec| vec.len());
+
+        // if we've traversed this block before for this peer (not necessarily for this particular request),
+        // there's no need to send it again
+        let block = data.filter(|_| self.link_tracker.block_ref_count(&link) == 0);
+        self.link_tracker
+            .record_link_traversal(id, link.clone(), block_is_present);
+
+        let builder = self.response_builder(block_size);
+        builder.add_link(id, link, block_is_present);
+
+        if let Some(block) = block {
+            builder.add_block(block);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Adds the given extension data to to the response.
+    pub fn send_extension_data(&mut self, id: RequestID, extension_data: ExtensionData) {
+        // we pass 0 as the block size since we're not adding any blocks to the response
+        self.response_builder(0)
+            .add_extension_data(id, extension_data);
+    }
+
+    /// Marks the given request ID as having sent all responses.
+    pub fn finish_request(&mut self, id: RequestID) -> ResponseStatusCode {
```

Comment:
can't really comment anywhere else but I assume just leaving the request erroring and pausing for when we need it later? or is there a specific reason leaving out?